boardswarm/src/hifive_p550_mcu.rs

```diff
@@ -0,0 +1,371 @@
+                    return Err(McuError::UnexpectedEof());
+                }
+                Err(e) => {
+                    return Err(e.into());
+                }
+            },
+        }
+    }
+}
+
+async fn process_som_power(
+    reader: &mut BufReader<ReadHalf<SerialStream>>,
+    write: &mut WriteHalf<SerialStream>,
+    parameters: SomPowerParameters,
+) -> Result<(), McuError> {
+    let (value, message) = parameters.mode.to_status();
+
+    {
+        let request = "sompower-g\n";
+        mcu_send(write, request).await?;
+    }
+
+    {
+        let response = Regex::new(r"^Som Power Status: (?<mode>.*)$").unwrap();
+        let ret = mcu_wait(reader, &response, MCU_RESPONSE_TIMEOUT_MILLIS).await?;
+        let caps = response.captures(&ret).unwrap();
+        if caps["mode"].to_lowercase() == parameters.mode.to_string() {
+            return Ok(());
+        }
+    }
+
+    {
+        let request = format!("sompower-s {}\n", value);
+        mcu_send(write, &request).await?;
+
+        let response = Regex::new(&message).unwrap();
+        let _ = mcu_wait(reader, &response, MCU_RESPONSE_TIMEOUT_MILLIS).await?;
+    }
+
+    Ok(())
+}
+
+async fn process_boot_sel(
+    reader: &mut BufReader<ReadHalf<SerialStream>>,
+    write: &mut WriteHalf<SerialStream>,
+    parameters: BootSelParameters,
+) -> Result<(), McuError> {
+    let (mode, value, bits) = parameters.mode.to_status();
+
+    {
+        let request = "bootsel-g\n";
+        mcu_send(write, request).await?;
+    }
+
+    {
+        let response = Regex::new(
+            r"^Get: Bootsel Controlled by: (?<mode>.*), bootsel\[3 2 1 0\]:(?<bits>.*)$",
+        )
+        .unwrap();
+        let ret = mcu_wait(reader, &response, MCU_RESPONSE_TIMEOUT_MILLIS).await?;
+        let caps = response.captures(&ret).unwrap();
+        if caps["mode"] == mode && bits.as_ref().is_some_and(|b| caps["bits"] == *b) {
+            return Ok(());
+        }
+    }
+
+    {
+        let request = format!("bootsel-s {} 0x{:x}\n", mode, value);
+        mcu_send(write, &request).await?;
+
+        let response = Regex::new(
+            r"^Set: Bootsel Controlled by: (?<mode>.*), bootsel\[3 2 1 0\]:(?<bits>.*)$",
+        )
+        .unwrap();
+        let ret = mcu_wait(reader, &response, MCU_RESPONSE_TIMEOUT_MILLIS).await?;
+        let caps = response.captures(&ret).unwrap();
+        if caps["mode"] == mode && bits.as_ref().is_some_and(|b| caps["bits"] == *b) {
+            return Ok(());
+        }
+    }
+
+    Ok(())
+}
+
+async fn process(
+    port: tokio_serial::SerialStream,
+    mut exec: tokio::sync::mpsc::Receiver<ChannelMessage>,
+) {
+    let (read, mut write) = tokio::io::split(port);
+    let mut reader = BufReader::new(read);
+
+    while let Some((command, parameters, resp)) = exec.recv().await {
+        let ret = match command {
+            McuCommand::HifiveP550McuSomPower => {
+                let parameters = SomPowerParameters::deserialize(parameters).unwrap();
```

Comment:
deserialize at the api entry point; not inside the process loop

Reply:
Fixed.